File: typescript/src/source.ts
import type { IconFetcher } from "./fetcher";
import type { IconResult } from "./result";

export type NetworkId = number;

export type Address = `0x${string}`;

const ADDRESS_PATTERN = /^0x[\da-f]{40}$/i;

export const isAddress = (value: string): value is Address => ADDRESS_PATTERN.test(value);

export type IconQuery
  = | { type: "network"; networkId: NetworkId; }
    | { type: "native"; networkId: NetworkId; }
    | { type: "erc20"; networkId: NetworkId; address: Address; };

export type IconSource = {
  name: string;
  url: (query: IconQuery) => string | undefined;
  fetch?: (fetcher: IconFetcher, query: IconQuery) => Promise<IconResult>;
};

File: cargo/src/source.rs
use crate::{
    Error, IconFetcher,
    identity::{Address, NetworkId},
    result::IconResult,
};

#[derive(Debug, Clone)]
pub enum IconQuery {
    // Ethereum Mainnet, Sepolia, etc.
    Network(NetworkId),
    // ETH, sepETH, etc.
    Native(NetworkId),
    // wETH, etc.
    ERC20(NetworkId, Address),
}

#[async_trait::async_trait]
pub trait IconSource: Send + Sync {
    fn name(&self) -> &'static str;

    fn url(&self, query: &IconQuery) -> Option<String>;

    async fn fetch(&self, fetcher: &IconFetcher, query: &IconQuery) -> Result<IconResult, Error> {
        match self.url(query) {
            Some(url) => fetcher.fetch_image_url(&url).await,
            None => Ok(IconResult::Unsupported),
        }
    }
}
